kore/src/Kore/Builtin/Set.hs

```diff
@@ -499,6 +589,631 @@ isSymbolUnit
+                        { term = NormalizedSet
+                            { elementsWithVariables = withVariableSet
+                            , concreteElements = concreteSet
+                            , sets = setsSet
+                            }
+                        , predicate
+                        , substitution
+                        }
+                    -- Add all unification predicates to the result.
+                    result =
+                        List.foldl'
+                            andCondition
+                            incompleteResult
+                            (  almostResultPredicates
+                            ++ setsPredicates
+                            )
+                return result
+  where
+    commonElements = Set.intersection concreteElements1 concreteElements2
+    commonVariables =
+        Set.intersection elementsWithVariables1 elementsWithVariables2
+    commonSets = Set.intersection sets1 sets2
+
+    elementDifference1 =
+        Set.toList (Set.difference concreteElements1 commonElements)
+    elementDifference2 =
+        Set.toList (Set.difference concreteElements2 commonElements)
+    elementVariableDifference1 =
+        Set.toList (Set.difference elementsWithVariables1 commonVariables)
+    elementVariableDifference2 =
+        Set.toList (Set.difference elementsWithVariables2 commonVariables)
+    setsDifference1 =
+        Set.toList (Set.difference sets1 commonSets)
+    setsDifference2 =
+        Set.toList (Set.difference sets2 commonSets)
+
+    errorForOpaqueSets =
+        (error . unlines)
+            [ "Unification case that should be handled somewhere else:"
+            , "attempting normalized unification with only an opaque"
+            , "set could lead to infinite loops."
+            , "first=" ++ unparseToString first
+            , "second=" ++ unparseToString second
+            ]
+
+    allSetElements1 =
+        map WithVariablePat elementVariableDifference1
+        ++ map (ConcretePat . TermLike.fromConcrete) elementDifference1
+    allSetElements2 =
+        map WithVariablePat elementVariableDifference2
+        ++ map (ConcretePat . TermLike.fromConcrete) elementDifference2
+
+    addAllDisjoint :: Ord a => Set a -> [a] -> unifier (Set a)
+    addAllDisjoint set elements =
+        case addToSetDisjoint set elements of
+            Nothing ->
+                Monad.Unify.explainAndReturnBottom
+                    "Duplicated elements after set unification."
+                    first
+                    second
+            Just result -> return result
+
+    simplify :: TermLike variable -> unifier (Pattern variable)
+    simplify term = alternate $ simplifyConditionalTerm term Predicate.top
+
+{- | Wrapper for terms that keeps the "concrete" vs "with variable" distinction
+after converting @TermLike Concrete@ to @TermLike variable@.
+-}
+data ConcreteOrWithVariable variable
+    = ConcretePat (TermLike variable)
```

Comment:
Let's leave it the way it is. You're right: it takes some time to traverse the term to do the transformation. (Generally, the compiler can't see through that transformation. There is a way to write the transformation so that the compiler can see through it... but I'm sure that's not worth our trouble at this point.)